File: App/Resources/AlgorithmDetails/combsort/combsort.swift
import Foundation

func sort(_ arr: inout [Int]) {
  let n = arr.count
  let shrink: Double = 1.3
  var gap = n
  var sorted = false
  while !sorted {
    gap = Int(floor(Double(gap) / shrink))
    if gap <= 1 {
      sorted = true
      gap = 1
    }
    for i in 0..<(n - gap) {
      let sm = gap + i
      if arr[i] > arr[sm] {
        arr.swapAt(i, sm)
        sorted = false
      }
    }
  }
}

var array: [Int] = [0, 39, 21, 62, 91, 77, 14, 23,
  90, 69, 51, 81, 68, 83, 32, 56]
sort(&array)
print(array)

File: App/Resources/AlgorithmDetails/combsort/combsort.c
#include <stdio.h>
#include <stdlib.h>

int array[16] = {0, 39, 21, 62, 91, 77, 14, 23,
                 90, 69, 51, 81, 68, 83, 32, 56};

void swap(int *a, int *b) {
  int t = *a;
  *a = *b;
  *b = t;
}

void printList(int items[], int size) {
  printf("[");
  if (size > 0) {
    printf("%d", items[0]);
    for (int i = 1; i < size; i++) {
      printf(", %d", items[i]);
    }
  }
  printf("]");
}

void sort(int arr[], int n) {
  int sm;
  float shrink = 1.3f;
  int gap = n;
  int sorted = 0;
  while (!sorted) {
    gap = (int)((float)gap / shrink);
    if (gap <= 1) {
      sorted = 1;
      gap = 1;
    }
    for (int i = 0; i < n - gap; i++) {
      sm = gap + i;
      if (arr[i] > arr[sm]) {
        swap(&arr[i], &arr[sm]);
        sorted = 0;
      }
    }
  }
}

int main(int argc, char *argv[]) {
  int size = sizeof(array) / sizeof(array[0]);
  sort(array, size);
  printList(array, size);
  return 0;
}
